### DISCUSSION TITLE
ckan-docker & harvest vs xloader approach to workers

### DISCUSSION THREAD

**Discussion by pdekraker-epa** [ckan/ckan#8066] [Q&A]
I am curious about best practice for installing ckanext-harvest within ckan-docker. I ask because I see a divergence between how harvest is mentioned in the ckan-docker [Dockerfile](https://github.com/ckan/ckan-docker/blob/89c900fe1c45565c6bb2723d99216ecf334ce44e/ckan/Dockerfile.dev#L20) and the [XLoader Wiki](https://github.com/ckan/ckan-docker/wiki/Replacing-DataPusher-with-XLoader). Both of these extensions require running separate workers. 

The source install of XLoader uses supervisor, but the recommendation is a separate container for docker. This approach makes sense to me as the logs would be more accessible and it is easier to manage with only a single item running in a container. A wrinkle for harvester is that it may require a harvester provided by another extension. This could be accommodated by running other containers utilizing the the same image (of ckan, harvester and all the desired extensions), just with different commands to start the workers instead. So I am wondering what is the best approach?

**Comment by thegostev** [comment]
FYI @kowh-ai

**Comment by kowh-ai** [comment]
Yes this is all very interesting, I think there are advantages and disadvantages for both running everything (UI +  extensions) in the one container or running UI and extensions (that run independently to the UI web server) in separate containers. Probably the biggest problem with multiple containers is “more containers = more maintenance” eg: when the CKAN code base changes. You are right, it would be cleaner / simpler to have separate containers for each extension. Perhaps a “worker container” is a better solution to both  these alternatives See: https://github.com/ckan/ckan-docker/issues/118

There is a decent chunk of work scheduled for the next 6 months to replace DataPusher with XLoader and then further down the track replacing XLoader with DataPusher+
